3rdparty/stout/tests/decompress_tests.cpp

```diff
@@ -0,0 +1,318 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License
+
+#include <string>
+
+#include <stout/base64.hpp>
+#include <stout/decompress.hpp>
+#include <stout/gtest.hpp>
+#include <stout/os.hpp>
+
+#include <stout/tests/utils.hpp>
+
+using std::string;
+
+
+class DecompressTest : public TemporaryDirectoryTest {};
+
+// No input file should return some error, not read from stdin.
+TEST_F(DecompressTest, DecompressEmptyInputFile)
+{
+  EXPECT_ERROR(decompress::extract("", ""));
+}
+
+// Non-existant file should return some error.
+TEST_F(DecompressTest, DecompressInputFileNotFound)
+{
+  // Construct a temporary file path that is guarenteed unique.
+  Try<string> dir = os::mkdtemp(path::join(os::getcwd(), "XXXXXX"));
+  ASSERT_SOME(dir);
+
+  string path = path::join(dir.get(), "ThisFileDoesNotExist");
+
+  EXPECT_ERROR(decompress::extract(path, ""));
+}
+
+TEST_F(DecompressTest, DecompressTarGzFile)
+{
+  // Construct a hello.tar.gz file that can be decompressed
+  string dir = path::join(os::getcwd(), "somedir");
+  ASSERT_SOME(os::mkdir(dir));
+
+  Try<string> path = os::mktemp(path::join(dir, "XXXXXX"));
+  ASSERT_SOME(path);
+
+  //  Length     Size     Date    Time  Name    Content
+  // --------  ------- ---------- ----- ----    ------
+  //       22       22 2018-02-21 10:06 hello   Howdy there, partner!\n
+  // --------  -------                  ------- ------
+  //       22       22                  1 file
+
+  ASSERT_SOME(os::write(path.get(), base64::decode(
+      "H4sICE61jVoAA2hlbGxvLnRhcgDtzjEOwjAQRNGtOcXSU9hx7FyBa0RgK0IR"
+      "RsYIcfsEaGhQqggh/VfsFLPFDHEcs6zLzEJon2k7bz7zrQliXdM6Nx/vxVgb"
+      "uiBqVt71crvWvqjKKaZ0yOnr31L/p/b5fnxoHWKJO730pZ5j2W5+vQoAAAAA"
+      "AAAAAAAAAAAAsGQC2DPIjgAoAAA=").get()));
+
+  // Note: The file does NOT have a .tar.gz extension. We could rename
+  // it, but libarchive doesn't care about extensions. It determines
+  // the format from the contents of the file. So this is tested here
+  // as well.
+  EXPECT_SOME(decompress::extract(path.get(), ""));
+
+  string extractedFile = path::join(os::getcwd(), "hello");
+  ASSERT_TRUE(os::exists(extractedFile));
+
+  ASSERT_SOME_EQ("Howdy there, partner!\n", os::read(extractedFile));
+}
+
+
+TEST_F(DecompressTest, DecompressTarFile)
```

Comment:
Add a similar test for a plain-text file to assert it errors (and then we can ask, should this abstraction detect such a potential error and avoid it, or return the error immediately, or just use `libarchive` and return its error?)

Reply:
I believe libarchive error should just be returned, I'd rather not be in the business of guessing what libarchive may or may not allow. Added a unit test for this.